packages/pinning-crust-backend/src/index.ts

```diff
@@ -0,0 +1,205 @@
+export class CrustPinningBackend implements PinningBackend {
+  static designator = 'crust'
+
+  readonly id: string
+  readonly endpoint: string
+  readonly seed: string
+  api: ApiPromise
+
+  constructor(readonly connectionString: string) {
+    // Check input
+    const url = new URL(connectionString)
+    const hostname = url.hostname
+    const seed = url.searchParams.get('seed')
+    if (!seed) {
+      throw new EmptySeedError(this.endpoint)
+    }
+    this.seed = seed
+
+    const protocol = url.protocol
+      .replace('crust:', 'ws')
+      .replace('crust+ws:', 'ws')
+      .replace('crust+wss:', 'wss')
+
+    if (url.port) {
+      this.endpoint = `${protocol}://${hostname}:${url.port}`
+    } else {
+      this.endpoint = `${protocol}://${hostname}`
+    }
+
+    // ID
+    const bytes = textEncoder.encode(this.connectionString)
+    const digest = toString(sha256.hash(bytes), 'base64urlpad')
+    this.id = `${CrustPinningBackend.designator}@${digest}`
+  }
+
+  open(): void {
+    this.api = new ApiPromise({
+      provider: new WsProvider(this.endpoint),
+      typesBundle: typesBundleForPolkadot,
+    });
+  }
+
+  async close(): Promise<void> {
+    if (this.api) {
+      await this.api.disconnect()
+      this.api = undefined
+    }
+  }
+
+  async pin(cid: CID): Promise<void> {
+    // Check connnetion
+    if (!this.api) {
+      return
+    }
+    await this.api.isReadyOrError
+
+    // Get a onchain keypair
+    const kr = new Keyring({ type: 'sr25519', ss58Format: 66 })
+    const krp = kr.addFromUri(this.seed)
+
+    // Storage order
+    // Support for the storage of up to 30G
+    const size = 30 * 1024 * 1024 * 1024
+    const tx = this.api.tx.market.placeStorageOrder(cid.toString(), size, 0, '')
+    await this.sendTx(tx, krp)
+
+    // Permanent storage
+    // Learn what's prepard for: https://wiki.crust.network/docs/en/DSM#3-file-order-assurance-settlement-and-discount
+    // In pCRU, 1 pCRU = 10^-12 CRU
+    const prepard = 1000000000 // in pCRU, 1 pCRU = 10^-12 CRU
+    const tx2 = this.api.tx.market.addPrepaid(cid.toString(), prepard)
+    await this.sendTx(tx2, krp)
+  }
+
+  async unpin(_cid: CID): Promise<void> {
+    // Do nothing
+  }
+
+  async ls(): Promise<CidList> {
+    // Check connnetion
+    if (!this.api) {
+      return {}
+    }
+    await this.api.isReadyOrError
+
+    // Get a onchain keypair
+    const kr = new Keyring({ type: 'sr25519', ss58Format: 66 })
+    const krp = kr.addFromUri(this.seed)
+    const data = '{"query": "query MyQuery {\\n  substrate_extrinsic(where: {method: {_eq: \\"placeStorageOrder\\"}, blockNumber: {}, signer: {_eq: \\"' + krp.address + '\\"}}, order_by: {blockNumber: desc}) {\\n    args(path: \\".[0].value\\")\\n  }\\n}\\n"}';
+
+    // Request
+    const result: CidList = {}
+    const tryTimes = 3
+    for (let i = 0; i < tryTimes; i++) {
+      try {
+        const res: AxiosResponse = await axios.post('https://crust.indexer.gc.subsquid.io/v4/graphql/', data)
```

Comment:
Please, use cross-fetch here instead of axios. One less library to load.

Also, is there a way to configure this endpoint maybe?

Reply:
I mean, otherwise it is ok to merge.

Reply:
Thanks for your comment, I have changed it to cross-fetch.

Reply:
And this endpoint is fixed and does not need to be changed